Fix crash when getting board.SPI outside the VM

tannewt opened the pull request:
If one of the default pins was already in use it would crash.

The internal API has been refined to allow us to get the value
without causing an init of the singleton.

Fixes #1753

ladyada commented:
tested and it worked on my cpx!

dhalbert commented on supervisor/shared/board_busses.h:
```diff
@@ -29,7 +29,8 @@
 
 #include "py/obj.h"
 
-mp_obj_t board_i2c(void);
```
The declaration of `board_i2c()` got deleted here, though `board_spi()` and `board_uart()` still exist. Was that deliberate?

tannewt replied:
Ya, I only switched one. Will push a larger refactor shortly.

dhalbert commented on shared-module/displayio/__init__.c:
```diff
@@ -190,7 +190,7 @@ void reset_displays(void) {
             ((uint32_t) fourwire->bus) > ((uint32_t) &displays + CIRCUITPY_DISPLAY_LIMIT)) {
             busio_spi_obj_t* original_spi = fourwire->bus;
-            if (original_spi == board_spi()) {
+            if (original_spi == common_hal_board_get_spi()) {
```
Could you put in a comment here about the fact the `original_spi` or `common_hal_board_get_spi()` might be null, and if so why, and how that's handled? I'm having trouble figuring out the different possible cases.

dhalbert commented on supervisor/shared/board_busses.c:
```diff
@@ -71,23 +71,42 @@ MP_DEFINE_CONST_FUN_OBJ_0(board_i2c_obj, board_i2c);
+    }
+    busio_spi_obj_t *self = &spi_obj;
+    self->base.type = &busio_spi_type;
+    if (!common_hal_mcu_pin_is_free(DEFAULT_SPI_BUS_SCK) ||
```
Could you add a comment? So, if the pins aren't free, we don't create, but just return null. Only `board_spi()` throws errors.

dhalbert commented on supervisor/shared/safe_mode.c:
```diff
@@ -77,7 +77,8 @@ safe_mode_t wait_for_safe_mode_reset(void) {
 
-void reset_into_safe_mode(safe_mode_t reason) {
+// Inline this so it's easy to break on it from GDB.
+void __attribute__((noinline,)) reset_into_safe_mode(safe_mode_t reason) {
```
"Don't inline this"?

tannewt commented:
Ok, I've refactored everything so it touches many more files. Hopefully it all makes more sense. I did add a `gc_never_free` which is a catch all linked list from a root pointer for pointers that should live the whole rest of the heap. (Like SPI objects used by FourWire.)

dhalbert commented:
> I did add a `gc_never_free` which is a catch all linked list from a root pointer for pointers that should live the whole rest of the heap. (Like SPI objects used by FourWire.)

That's a great idea!

dhalbert commented on shared-module/board/__init__.c:
```diff
@@ -0,0 +1,115 @@
+STATIC busio_spi_obj_t spi_obj;
+STATIC mp_obj_t spi_singleton = NULL;
+
+// TODO(tannewt): Move this to shared-bindings/board/__init__.c and corresponding shared-module.
```
You can remove this TODO!

dhalbert commented:
Bunch of builds failing :(

tannewt commented:
Ok, I think this next run should pass. I fixed boards without shared busses, removed `audiobusio` from hallowing and removed `touchio` from pirkey to gain space back.
